File: scripts/check-staged-coverage.mjs
#!/usr/bin/env node
// Per-file coverage gate. Reads coverage/coverage-summary.json (vitest /
// v8 / istanbul output) and verifies each staged .ts/.tsx file meets the
// thresholds. Used by setup-ts-flavor's pre-commit hook.
//
// Behavior:
//   - No staged .ts/.tsx files                → exit 0 (nothing to check)
//   - coverage-summary.json missing           → exit 0 with a one-line hint
//                                               (so bootstrap commits aren't
//                                               blocked before tests exist)
//   - Staged file present in report + below   → BLOCK (exit 1)
//   - Staged file absent from report          → BLOCK (exit 1) — staged
//                                               source code with no test
//                                               file at all is the worst
//                                               quality signal we have
//
// Override (escape hatch for legitimate bootstrap commits with no tests):
//   STAGED_COVERAGE_SKIP=1 git commit ...

import { execSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';

const THRESHOLDS = {
  statements: 75,
  branches: 60,
  functions: 75,
  lines: 75,
};

if (process.env.STAGED_COVERAGE_SKIP === '1') {
  console.log('STAGED_COVERAGE_SKIP=1 — coverage gate bypassed (use sparingly).');
  process.exit(0);
}

const stagedFiles = execSync('git diff --cached --name-only --diff-filter=ACM', { encoding: 'utf-8' })
  .split('\n')
  .filter((f) => /\.(ts|tsx)$/.test(f))
  .filter((f) => !/\.(test|spec)\.tsx?$/.test(f))
  .filter(Boolean);

if (stagedFiles.length === 0) {
  console.log('Coverage gate: no staged .ts/.tsx — skipped.');
  process.exit(0);
}

const coverageFile = path.join(process.cwd(), 'coverage', 'coverage-summary.json');
if (!fs.existsSync(coverageFile)) {
  console.log(
    'Coverage gate: coverage/coverage-summary.json not found — skipped.\n' +
      '  Run `npm run test:coverage` to enable the gate. Once a report exists,\n' +
      '  this gate will enforce per-file thresholds on subsequent commits.',
  );
  process.exit(0);
}

const coverage = JSON.parse(fs.readFileSync(coverageFile, 'utf-8'));

const failures = [];
const notFound = [];

for (const file of stagedFiles) {
  const absolutePath = path.resolve(process.cwd(), file);
  const fileCoverage = coverage[absolutePath];

  if (!fileCoverage) {
    notFound.push(file);
    continue;
  }

  const issues = [];
  for (const [metric, threshold] of Object.entries(THRESHOLDS)) {
    const pct = fileCoverage[metric]?.pct ?? 0;
    if (pct < threshold) issues.push(`${metric}: ${pct.toFixed(2)}% < ${threshold}%`);
  }
  if (issues.length > 0) failures.push({ file, issues });
}

if (failures.length === 0 && notFound.length === 0) {
  console.log(`Coverage gate: ${stagedFiles.length} staged file(s) above thresholds. ✓`);
  process.exit(0);
}

console.error('');
console.error('╔══════════════════════════════════════════════════════════════════╗');
console.error('║  COVERAGE GATE FAILED                                            ║');
console.error('╚══════════════════════════════════════════════════════════════════╝');

if (failures.length > 0) {
  console.error('');
  console.error('Below threshold:');
  for (const { file, issues } of failures) {
    console.error(`  ${file}`);
    for (const issue of issues) console.error(`    ${issue}`);
  }
}

if (notFound.length > 0) {
  console.error('');
  console.error('No coverage report entry (file has no associated test):');
  for (const file of notFound) console.error(`  ${file}`);
}

console.error('');
console.error('Thresholds:', JSON.stringify(THRESHOLDS));
console.error('Fix: add or improve tests, then `npm run test:coverage` to refresh the report.');
console.error('Bypass (use sparingly):  STAGED_COVERAGE_SKIP=1 git commit ...');
console.error('');
process.exit(1);

File: scripts/check-staged-coverage.py
#!/usr/bin/env python3
"""Per-file coverage gate for Python (parallel to check-staged-coverage.mjs).

Reads `coverage/coverage-summary.json` produced by `coverage json` (after
pytest-cov). Verifies each staged .py file (excluding tests/) meets per-file
thresholds.

Behavior matches the TS variant:
  - No staged .py files                       -> exit 0 (nothing to check)
  - coverage/coverage-summary.json missing    -> exit 0 with hint (bootstrap-friendly)
  - Staged file present + below threshold     -> BLOCK (exit 1)
  - Staged file absent from report            -> BLOCK (exit 1) (source with no test)

Override: STAGED_COVERAGE_SKIP=1 git commit ...
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

THRESHOLDS = {
    "percent_covered": 75.0,  # coverage.py reports overall %; statements + branches together
}
BRANCH_THRESHOLD = 60.0  # only checked if `branches` key present (requires --cov-branch)


def main() -> int:
    if os.environ.get("STAGED_COVERAGE_SKIP") == "1":
        print("STAGED_COVERAGE_SKIP=1 — coverage gate bypassed (use sparingly).")
        return 0

    staged = subprocess.run(
        ["git", "diff", "--cached", "--name-only", "--diff-filter=ACM"],
        capture_output=True, text=True, check=True,
    ).stdout.splitlines()

    py_files = [
        f for f in staged
        if f.endswith(".py")
        and not f.startswith("tests/")
        and "/tests/" not in f
        and not Path(f).name.startswith("test_")
        and not Path(f).name.endswith("_test.py")
    ]

    if not py_files:
        print("Coverage gate: no staged .py — skipped.")
        return 0

    report_path = Path.cwd() / "coverage" / "coverage-summary.json"
    if not report_path.exists():
        print(
            "Coverage gate: coverage/coverage-summary.json not found — skipped.\n"
            "  Run `pytest --cov-report=json:coverage/coverage-summary.json` to enable\n"
            "  the gate. Once a report exists, this gate enforces thresholds on\n"
            "  subsequent commits.",
        )
        return 0

    report = json.loads(report_path.read_text())
    files_section = report.get("files", {})

    failures: list[tuple[str, list[str]]] = []
    not_found: list[str] = []

    for staged_file in py_files:
        entry = files_section.get(staged_file)
        if entry is None:
            not_found.append(staged_file)
            continue

        summary = entry.get("summary", {})
        issues: list[str] = []

        pct = summary.get("percent_covered")
        if pct is not None and pct < THRESHOLDS["percent_covered"]:
            issues.append(f"percent_covered: {pct:.2f}% < {THRESHOLDS['percent_covered']}%")

        branch_pct = summary.get("percent_covered_branches")
        if branch_pct is not None and branch_pct < BRANCH_THRESHOLD:
            issues.append(f"branches: {branch_pct:.2f}% < {BRANCH_THRESHOLD}%")

        if issues:
            failures.append((staged_file, issues))

    if not failures and not not_found:
        print(f"Coverage gate: {len(py_files)} staged file(s) above thresholds. ✓")
        return 0

    print("", file=sys.stderr)
    print("╔══════════════════════════════════════════════════════════════════╗", file=sys.stderr)
    print("║  COVERAGE GATE FAILED                                            ║", file=sys.stderr)
    print("╚══════════════════════════════════════════════════════════════════╝", file=sys.stderr)

    if failures:
        print("\nBelow threshold:", file=sys.stderr)
        for f, issues in failures:
            print(f"  {f}", file=sys.stderr)
            for issue in issues:
                print(f"    {issue}", file=sys.stderr)

    if not_found:
        print("\nNo coverage report entry (file has no associated test):", file=sys.stderr)
        for f in not_found:
            print(f"  {f}", file=sys.stderr)

    print(
        f"\nThresholds: percent_covered >= {THRESHOLDS['percent_covered']}%, "
        f"branches >= {BRANCH_THRESHOLD}% (if --cov-branch)",
        file=sys.stderr,
    )
    print(
        "Fix: add or improve tests, then "
        "`pytest --cov-report=json:coverage/coverage-summary.json` to refresh.",
        file=sys.stderr,
    )
    print("Bypass (use sparingly):  STAGED_COVERAGE_SKIP=1 git commit ...", file=sys.stderr)
    print("", file=sys.stderr)
    return 1


if __name__ == "__main__":
    sys.exit(main())
